Test failure on 32bit linux looks real. Going to restart out of curiosity if it is reproducible (since I have currently no idea why it would fail).

<details> <summary> Failure </summary>

```
=================================== FAILURES ===================================
_______ TestNewBufferProtocol.test_export_and_pickle_user_dtype[scalar] ________

self = <numpy.core.tests.test_multiarray.TestNewBufferProtocol object at 0xec9d1238>
obj = rational(1,2), error = <class 'TypeError'>

    @pytest.mark.parametrize(["obj", "error"], [
            pytest.param(np.array([1, 2], dtype=rational), ValueError, id="array"),
            pytest.param(rational(1, 2), TypeError, id="scalar")])
    def test_export_and_pickle_user_dtype(self, obj, error):
        # User dtypes should export successfully when FORMAT was not requested.
        with pytest.raises(error):
            _multiarray_tests.get_buffer_info(obj, ("STRIDED", "FORMAT"))
    
        _multiarray_tests.get_buffer_info(obj, ("STRIDED",))
    
        # This is currently also necessary to implement pickling:
        res = pickle.loads(pickle.dumps(obj))
>       assert_array_equal(res, obj)
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference: rational(255,514)
E       Max relative difference: rational(255,257)
E        x: array(rational(1,257), dtype=rational)
E        y: array(rational(1,2), dtype=rational)

error      = <class 'TypeError'>
obj        = rational(1,2)
res        = rational(1,257)
self       = <numpy.core.tests.test_multiarray.TestNewBufferProtocol object at 0xec9d1238>

../../venv/lib/python3.8/site-packages/numpy/core/tests/test_multiarray.py:7159: AssertionError
```

--- a/numpy/core/src/multiarray/buffer.c
+++ b/numpy/core/src/multiarray/buffer.c
@@ -637,6 +637,8 @@ _buffer_get_info(PyObject *obj, int flags)
             item = PyList_GetItem(item_list, item_list_length - 1);
             old_info = (_buffer_info_t*)PyLong_AsVoidPtr(item);
             if (_buffer_info_cmp(info, old_info) != 0) {
+                old_info = NULL;  /* Can't use this one, but possibly next */
+
                 if (item_list_length > 1 && info->ndim > 1) {
                     /*
                      * Some arrays are C- and F-contiguous and if they have more
@@ -653,9 +655,6 @@ _buffer_get_info(PyObject *obj, int flags)
                     if (_buffer_info_cmp(info, old_info) != 0) {
                         old_info = NULL;
                     }
-                }
-                else {
-                    old_info = NULL;
                 }
             }
 

--- a/numpy/core/src/multiarray/scalarapi.c
+++ b/numpy/core/src/multiarray/scalarapi.c
@@ -35,7 +35,7 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
 {
     int type_num;
     int align;
-    npy_intp memloc;
+    uintptr_t memloc;
     if (descr == NULL) {
         descr = PyArray_DescrFromScalar(scalar);
         type_num = descr->type_num;
@@ -168,7 +168,7 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
      * Use the alignment flag to figure out where the data begins
      * after a PyObject_HEAD
      */
-    memloc = (npy_intp)scalar;
+    memloc = (uintptr_t)scalar;
     memloc += sizeof(PyObject);
     /* now round-up to the nearest alignment value */
     align = descr->alignment;

--- a/numpy/core/tests/test_multiarray.py
+++ b/numpy/core/tests/test_multiarray.py
@@ -7155,7 +7155,8 @@ class TestNewBufferProtocol:
         _multiarray_tests.get_buffer_info(obj, ("STRIDED",))
 
         # This is currently also necessary to implement pickling:
-        res = pickle.loads(pickle.dumps(obj))
+        pickle_obj = pickle.dumps(obj)
+        res = pickle.loads(pickle_obj)
         assert_array_equal(res, obj)
 
     def test_padding(self):

